h5py/_hl/vds.py

```diff
@@ -156,17 +160,83 @@ class VirtualLayout(object):
     maxshape
         The virtual dataset is resizable up to this shape. Use None for
         axes you want to be unlimited.
+    filename
+        The name of the destination file, if known in advance. Mappings from
+        data in the same file will be stored with filename '.', allowing the
+        file to be renamed later.
     """
-    def __init__(self, shape, dtype, maxshape=None):
+    def __init__(self, shape, dtype, maxshape=None, filename=None):
         self.shape = (shape,) if isinstance(shape, int) else shape
         self.dtype = dtype
         self.maxshape = (maxshape,) if isinstance(maxshape, int) else maxshape
-        self.sources = []
+        self.filename = filename
+        self._src_filenames = set()
+        self.dcpl = h5p.create(h5p.DATASET_CREATE)
 
     def __setitem__(self, key, source):
         sel = select(self.shape, key, dataset=None)
         _convert_space_for_key(sel.id, key)
-        self.sources.append(VDSmap(sel.id,
-                                   source.path,
-                                   source.name,
-                                   source.sel.id))
+        src_filename = self._source_file_name(source.path, self.filename)
+
+        self.dcpl.set_virtual(
+            sel.id, src_filename, source.name.encode('utf-8'), source.sel.id
+        )
+        if self.filename is None:
+            self._src_filenames.add(src_filename)
+
+    @staticmethod
+    def _source_file_name(src_filename, dst_filename) -> bytes:
+        src_filename = filename_encode(src_filename)
+        if dst_filename and (src_filename == filename_encode(dst_filename)):
+            # use relative path if the source dataset is in the same
+            # file, in order to keep the virtual dataset valid in case
+            # the file is renamed.
+            return b'.'
+        return filename_encode(src_filename)
+
+    def get_dcpl(self, dst_filename=None):
```

Comment:
Should `dst_filename` be required?  If `None` makes in into `filename_encode` it will explode with 

```
In [10]: filename_encode(None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-eade088af214> in <module>
----> 1 filename_encode(None)

~/.virtualenvs/sys39/lib/python3.9/site-packages/h5py/_hl/compat.py in filename_encode(filename)
     17     filenames in h5py for more information.
     18     """
---> 19     filename = fspath(filename)
     20     if sys.platform == "win32":
     21         if isinstance(filename, str):

TypeError: expected str, bytes or os.PathLike object, not NoneType
```

Reply:
Ah, yes, you're right. I was changing my mind about whether to require it or not.